sql: add a retry loop for stmts in READ COMMITTED txns

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/100145

This new loop retries individual statements inside an explicit READ
COMMITTED transaction. This is possible because each statement in a READ
COMMITTED transaction has a different read timestamp. The conn executor
already has retry logic for all implicit transactions, and we continue to
use those when possible.

A session setting controls how many retries will be performed for a statement
inside of an explicit READ COMMITTED transaction.

Release note (sql change): Added the
max_retries_for_read_committed session variable. It
defaults to 10, and determines the number of times an individual
statement in an explicit READ COMMITTED transaction will be retried
if it encounters a retriable transaction error.

rafiss committed:
sql: add a retry loop for stmts in READ COMMITTED txns

This new loop retries individual statements inside an explicit READ
COMMITTED transaction. This is possible because each statement in a READ
COMMITTED transaction has a different read timestamp. The conn executor
already has retry logic for all implicit transactions, and we continue to
use those when possible.

A session setting controls how many retries will be performed for a statement
inside of an explicit READ COMMITTED transaction.

Release note (sql change): Added the
max_retries_for_read_committed session variable. It
defaults to 10, and determines the number of times an individual
statement in an explicit READ COMMITTED transaction will be retried
if it encounters a retriable transaction error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107044)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> minus CI not being happy

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-:-N_e8OczBmQTYeEU_A-s:b-bkanft)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/vars.go:
```diff
@@ -2028,6 +2028,36 @@ var varGen = map[string]sessionVar{
+	// CockroachDB extension. Configures the maximum number of automatic retries
+	// to perform for statements in explicit READ COMMITTED transactions that
+	// see a transaction retry error.
+	`max_retries_for_read_committed_transactions`: {
```
how about a cluster setting for this purpose?

rafiss replied:
i lean towards session variable, since cluster settings are generally intended for usage by operators, and require higher permissions on the cluster. also, in this case, it is reasonable that different workloads may want different values here, since it depends on how much contention they see.

Xiang-Gu reviewed (approved):
Good work!

michae2 reviewed (commented):
Nice work!

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-:-N_onqP6DDXIP9e7Biz6:bdtx8r6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @fqazi, @nvanbenschoten, and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_onON_-4o25fLI6kmx:-N_onON_-4o25fLI6kmy:b-nmmo4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L963)):*
> ```Go
> 	if ex.state.mu.txn.IsOpen() &&
> 		ex.state.mu.txn.IsoLevel() == isolation.ReadCommitted &&
> 		!ex.implicitTxn() &&
> ```

Do we also want to defer to the state machine retry for the first statement of explicit transactions?
___
*[`pkg/sql/conn_executor_exec.go` line 980 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_om6Tr9u1upjb3lguj:-N_om6Tr9u1upjb3lguk:bzh39va) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L980)):*
> ```Go
> 	}
> 
> 	maxExecCount := 1
> ```

Rather than going through this loop codepath in every case, I think it would be cleaner to put the retry loop in a separate conn executor function that wraps `dispatchToExecutionEngine` and is only called for read committed transactions. That will do two things: (a) get rid of some of the conditional logic, and (b) add a function to the stack that will be visible when debugging which could help us know we're in a read committed transaction.
___
*[`pkg/sql/conn_executor_exec.go` line 987 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_okxTN77wiBN1kA-Gf:-N_okxTN77wiBN1kA-Gg:boqbyl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L987)):*
> ```Go
> 	}
> 
> 	for attemptNum := 0; attemptNum < maxExecCount; attemptNum++ {
> ```

Does this loop need some kind of (randomized) backoff? For example, I think two update statements could both conflict with each other and both need to retry repeatedly.
___
*[`pkg/sql/conn_executor_exec.go` line 1005 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_omvXD8wYHY6B7YKCM:-N_omvXD8wYHY6B7YKCN:b-k8vw6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L1005)):*
> ```Go
> 				break
> 			}
> 			if !errIsRetriable(maybeRetriableErr) || attemptNum == maxRetries {
> ```

Does `errIsRetriable` return true for any errors that need a full transaction retry (rather than just a statement retry)? If so, we might be performing extra useless statement retries in those cases, when we should go straight to a transaction-level retry.
___
*[`pkg/sql/vars.go` line 2034 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107044#gh-1267330389:-N_oem7y5LVIb0NkqSxL:bt2i69x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c309844935d4f6a8e850f4af307fb62deda9836/pkg/sql/vars.go#L2034)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i lean towards session variable, since cluster settings are generally intended for usage by operators, and require higher permissions on the cluster. also, in this case, it is reasonable that different workloads may want different values here, since it depends on how much contention they see.
</blockquote></details>

bikeshedding: I understand that "_transactions" refers to the fact that the whole transaction is at read-committed isolation, but find it a little confusing for this setting. I think it would be clearer to leave it off or change it to "_statements".


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @fqazi, and @nvanbenschoten)

___
*[`pkg/sql/conn_executor_exec.go` line 1019 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NaDX4Ag0FIlu_6ZmrDv:-NaDX4Ag0FIlu_6ZmrDw:b-305pzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c309844935d4f6a8e850f4af307fb62deda9836/pkg/sql/conn_executor_exec.go#L1019)):*
> ```Go
> 			}
> 			res.SetError(nil)
> 			ex.state.mu.txn.PrepareForRetry(ctx)
> ```

we should wait for [kv: remove TxnCoordSender.PrepareRetryableError, rationalize ManualRestart](https://github.com/cockroachdb/cockroach/pull/105109) to complete.


<!-- Sent from Reviewable.io -->

rafiss commented:
For some reason I can't type into Reviewable, but replying here:

> Do we also want to defer to the state machine retry for the first statement of explicit transactions?

I don't think so - the state machine doesn't do retries if the `BEGIN;` and first statement are sent in two different commands. But in this case, we would want to.

> Rather than going through this loop codepath in every case, I think it would be cleaner to put the retry loop in a separate conn executor function that wraps dispatchToExecutionEngine and is only called for read committed transactions. That will do two things: (a) get rid of some of the conditional logic, and (b) add a function to the stack that will be visible when debugging which could help us know we're in a read committed transaction.

Very great idea. Done, and hopefully it's simpler to read now.

> Does errIsRetriable return true for any errors that need a full transaction retry (rather than just a statement retry)? If so, we might be performing extra useless statement retries in those cases, when we should go straight to a transaction-level retry.

Great catch. I've now rebased on top of https://github.com/cockroachdb/cockroach/pull/105161/ so that I can check with `TxnMustRestartFromBeginning`.

> bikeshedding: I understand that "_transactions" refers to the fact that the whole transaction is at read-committed isolation, but find it a little confusing for this setting. I think it would be clearer to leave it off or change it to "_statements".

I just removed "_transactions" from the end now.

rafiss committed:
sql/tests: add a test for WriteTooOldError under READ COMMITTED

Release note: None

nvb reviewed (commented):
> Reviewed 1 of 1 files at r12, 3 of 3 files at r13, 11 of 11 files at r14, 5 of 5 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @rafiss, and @rhu713)

___
*[`-- commits` line 38 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbteqO-0eHmUNGwbLqU):*
s/Ratehr/Rather/
___
*[`pkg/sql/conn_executor_exec.go` line 1048 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbthXjr28io77n6BdTs:-NbthXjr28io77n6BdTt:b-8ljrg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1048)):*
> ```Go
> 	}
> 
> 	if ex.state.mu.txn.IsOpen() &&
> ```

This is called by `execStmtInOpenState`. Is it possible for the txn to not be open at this point?
___
*[`pkg/sql/conn_executor_exec.go` line 1059 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtiBaj1tXBBLF3YTJ0:-NbtiBaj1tXBBLF3YTJ1:b9qr98b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1059)):*
> ```Go
> 		// machine, we use the KV savepoint API directly.
> 		//
> 		// TODO(rafi): Find a way to perform retries in the internal executor
> ```

If an internal executor query that is run as part of a larger statement throws a retryable error, will that be returned up and retried by the statement's `dispatchReadCommittedStmtToExecutionEngine` retry loop?

It feels like (from here and https://github.com/cockroachdb/cockroach/issues/86162) we might be converging on two different forms of internal executor types, one that is performed on behalf of SQL in an executor with a larger scope, and one that is performed on behalf of a KV txn 
___
*[`pkg/sql/conn_executor_exec.go` line 1063 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkCG-29dA_EcF_0cb:-NbtkCG-29dA_EcF_0cc:b-d1mgcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1063)):*
> ```Go
> 		if convertErrToEvent, err := ex.dispatchReadCommittedStmtToExecutionEngine(stmtCtx, p, res); err != nil {
> 			stmtThresholdSpan.Finish()
> 			if convertErrToEvent {
> ```

Instead of returning `convertErrToEvent` and using that to decide what to do with the error, would it have been cleaner to return the full `(fsm.Event, fsm.EventPayload, error)` from `dispatchReadCommittedStmtToExecutionEngine` and let it call `ex.makeErrEvent` when desirable? My reading (could be incorrect) is that `p.stmt.AST` is available for `makeErrEvent`.
___
*[`pkg/sql/conn_executor_exec.go` line 1514 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkuoxBilZ5wmqkHpw:-NbtkuoxBilZ5wmqkHpx:bhm410w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1514)):*
> ```Go
> }
> 
> func (ex *connExecutor) dispatchReadCommittedStmtToExecutionEngine(
> ```

nit: I assume we'll want to add some commentary on this function about what it does and what role it plays.
___
*[`pkg/sql/conn_executor_exec.go` line 1523 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nbtl0PG0taTWWoDzq2l:-Nbtl0PG0taTWWoDzq2m:bblokon) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1523)):*
> ```Go
> 
> 	maxRetries := int(ex.sessionData().MaxRetriesForReadCommitted)
> 	for attemptNum := 0; ; attemptNum++ {
> ```

Have you given any thought to the kinds of observability we might want to provide for these retries?
___
*[`pkg/sql/conn_executor_exec.go` line 1538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtlMsXBvVEvzU3etBi:-NbtlMsXBvVEvzU3etBj:be2ngji) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1538)):*
> ```Go
> 		// If we reached the maximum number of retries, then we must stop.
> 		if attemptNum == maxRetries {
> 			break
> ```

Should we be returning an error in this case?
___
*[`pkg/sql/conn_executor_exec.go` line 1538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnwkK8-r0a6c-_ZYK:-NbtnwkK8-r0a6c-_ZYL:b-t7hfwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1538)):*
> ```Go
> 		// If we reached the maximum number of retries, then we must stop.
> 		if attemptNum == maxRetries {
> 			break
> ```

If this is a retry error, should we wrap it with some mention of the retry exhaustion? That might motivate moving this check down below the next one.
___
*[`pkg/sql/conn_executor_exec.go` line 1543 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtlTJbDlbcA1Xr6MLr:-NbtlTJbDlbcA1Xr6MLs:b-bfdyy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1543)):*
> ```Go
> 		var txnRetryErr *kvpb.TransactionRetryWithProtoRefreshError
> 		if !errors.As(maybeRetriableErr, &txnRetryErr) || txnRetryErr.TxnMustRestartFromBeginning() {
> 			break
> ```

Same question here, and also in the `break` below. I wonder if this logic was assigning `retErr = res.Err()` before and then it changed.
___
*[`pkg/sql/conn_io.go` line 849 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtmlBV-BVy4hR4Oayz:-NbtmlBV-BVy4hR4Oaz-:bdw1zmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_io.go#L849)):*
> ```Go
> 
> 	// BufferedResultsLen returns the length of the results buffer.
> 	BufferedResultsLen() int
> ```

Should this return a `CmdPos`? And should `TruncateBufferedResults` accept one?
___
*[`pkg/sql/conn_io.go` line 852 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtmCInAw2VzT2O3LgJ:-NbtmCInAw2VzT2O3LgK:b6iyjzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_io.go#L852)):*
> ```Go
> 
> 	// TruncateBufferedResults clears any results that have been buffered after
> 	// given position, and returns true iff any results were actually truncated.
> ```

If no results were truncated because no results were added since the call to `BufferedResultsLen`, do we want to return `true` or `false`? In other words, what should `TruncateBufferedResults(BufferedResultsLen())` return?

Given the use case for this method, I'd think we would want it to return true, but I think it will currently return false.
___
*[`pkg/sql/pgwire/command_result.go` line 273 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnPYLDdTrNlebR-1j:-NbtnPYLDdTrNlebR-1k:bnkd634) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/pgwire/command_result.go#L273)):*
> ```Go
> 
> // TruncateBufferedResults is part of the sql.RestrictedCommandResult interface.
> func (r *commandResult) TruncateBufferedResults(pos int) bool {
> ```

Should we add a unit test for this?
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 5351 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnUglBB2c26Is1F98:-NbtnUglBB2c26Is1F99:b-xz1m6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/logictest/testdata/logic_test/information_schema#L5351)):*
> ```
> max_identifier_length                                      128
> max_index_keys                                             32
> max_retries_for_read_committed                10
> ```

nit: the indentation looks off here and in a few other logic test files. Does `--rewrite` fix that?
___
*[`pkg/sql/tests/read_committed_test.go` line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtoUR9DZUE_oa2TXCj:-NbtoUR9DZUE_oa2TXCk:bxn0bgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1895f45d440541bbd1d4134afa35c39ef7131b/pkg/sql/tests/read_committed_test.go#L41)):*
> ```Go
> 
> 	var readCommittedWriteCount int
> 	var waitBeforeSerializableWriteToC sync.WaitGroup
> ```

nit: what does "serializable" refer to here?
___
*[`pkg/sql/tests/read_committed_test.go` line 76 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtofEiEg2SjghR9-GE:-NbtofEiEg2SjghR9-GF:b-bjydjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1895f45d440541bbd1d4134afa35c39ef7131b/pkg/sql/tests/read_committed_test.go#L76)):*
> ```Go
> 	defer s.Stopper().Stop(ctx)
> 
> 	// Create a table with three rows.
> ```

Consider calling out here that `k` is not the primary key, so locking won't be pushed into the initial scan of UPDATE statements.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @nvanbenschoten, and @rhu713)

___
*[`-- commits` line 38 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbteqO-0eHmUNGwbLqU):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/Ratehr/Rather/
</blockquote></details>

done
___
*[`pkg/sql/conn_executor_exec.go` line 963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_onON_-4o25fLI6kmx:-NbuES2j5OLkSRjKN9LU:b-vxj8n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L963)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we also want to defer to the state machine retry for the first statement of explicit transactions?
</blockquote></details>

I don't think so - the state machine doesn't do retries if the `BEGIN;` and first statement are sent in two different commands. But in this case, we would want to.
___
*[`pkg/sql/conn_executor_exec.go` line 980 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_om6Tr9u1upjb3lguj:-NbuEQlv81weFlvhCsJf:b-xh7tho) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L980)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Rather than going through this loop codepath in every case, I think it would be cleaner to put the retry loop in a separate conn executor function that wraps `dispatchToExecutionEngine` and is only called for read committed transactions. That will do two things: (a) get rid of some of the conditional logic, and (b) add a function to the stack that will be visible when debugging which could help us know we're in a read committed transaction.
</blockquote></details>

Very great idea. Done, and hopefully it's simpler to read now.
___
*[`pkg/sql/conn_executor_exec.go` line 987 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_okxTN77wiBN1kA-Gf:-NbuEFoBAKe-QzONo_CY:bo0k8wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L987)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does this loop need some kind of (randomized) backoff? For example, I think two update statements could both conflict with each other and both need to retry repeatedly.
</blockquote></details>

i don't think so - during this loop locks are still being held, so we shouldn't stay idle while holding them.
___
*[`pkg/sql/conn_executor_exec.go` line 1005 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-N_omvXD8wYHY6B7YKCM:-NbuFKkU0kt8aTGdjm49:bsld0me) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cffb7e200e85c81050eb39c69c3fe6430162e3e/pkg/sql/conn_executor_exec.go#L1005)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does `errIsRetriable` return true for any errors that need a full transaction retry (rather than just a statement retry)? If so, we might be performing extra useless statement retries in those cases, when we should go straight to a transaction-level retry.
</blockquote></details>

Great catch. I've now rebased on top of [#105161](https://github.com/cockroachdb/cockroach/pull/105161) so that I can check with `TxnMustRestartFromBeginning`.
___
*[`pkg/sql/conn_executor_exec.go` line 1048 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbthXjr28io77n6BdTs:-NbuETQhDtb2Bt0MVUj5:broecq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1048)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is called by `execStmtInOpenState`. Is it possible for the txn to not be open at this point?
</blockquote></details>

no it's not - i think this was left over from when i was figuring out where to do the retries. removed now
___
*[`pkg/sql/conn_executor_exec.go` line 1059 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtiBaj1tXBBLF3YTJ0:-NbuEk1j49uYYU2BT-SC:b-dd04my) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1059)):*
> If an internal executor query that is run as part of a larger statement throws a retryable error, will that be returned up and retried by the statement's `dispatchReadCommittedStmtToExecutionEngine` retry loop?

that does feel like what i'd expect to happen

> It feels like (from here and [![](https://avatars.githubusercontent.com/u/1839234?s=60&v=4) sql: internal executor usage bumps txn sequence number #86162](https://github.com/cockroachdb/cockroach/issues/86162)) we might be converging on two different forms of internal executor types, one that is performed on behalf of SQL in an executor with a larger scope, and one that is performed on behalf of a KV txn

Yeah - I think the `fromOuterTxn` flag of `extraTxnState` helps with this.
___
*[`pkg/sql/conn_executor_exec.go` line 1063 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkCG-29dA_EcF_0cb:-NbuF1liCm5OefzTvaV6:b-ks0r6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1063)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of returning `convertErrToEvent` and using that to decide what to do with the error, would it have been cleaner to return the full `(fsm.Event, fsm.EventPayload, error)` from `dispatchReadCommittedStmtToExecutionEngine` and let it call `ex.makeErrEvent` when desirable? My reading (could be incorrect) is that `p.stmt.AST` is available for `makeErrEvent`.
</blockquote></details>

yeah i'd considered doing it this way, and i'm fine with either approach
___
*[`pkg/sql/conn_executor_exec.go` line 1514 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkuoxBilZ5wmqkHpw:-NbuFLyx5nX8aFSI4c3V:b-a4j0xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1514)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I assume we'll want to add some commentary on this function about what it does and what role it plays.
</blockquote></details>

yeah good call
___
*[`pkg/sql/conn_executor_exec.go` line 1523 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nbtl0PG0taTWWoDzq2l:-NbuFOS67_oatxk6mX_l:b89mi6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1523)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Have you given any thought to the kinds of observability we might want to provide for these retries?
</blockquote></details>

one thought i had is if it would make sense to reuse the `autoRetryCounter` that we already use the for the conn_executor state machine retries. it would be a slight abuse of the variable, but it actually doesn't seem incorrect to me, and it would be the quickest way to plug into any existing observability that shows retries.
___
*[`pkg/sql/conn_executor_exec.go` line 1538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtlMsXBvVEvzU3etBi:-NbuVSZJ2kXog-Azlk8E:bj6lc6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1538)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be returning an error in this case?
</blockquote></details>

adding a comment
___
*[`pkg/sql/conn_executor_exec.go` line 1538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnwkK8-r0a6c-_ZYK:-NbuVUJpAVfxS_AfKvdB:bb4pus3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1538)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If this is a retry error, should we wrap it with some mention of the retry exhaustion? That might motivate moving this check down below the next one.
</blockquote></details>

wrapping it is a good idea.
___
*[`pkg/sql/conn_executor_exec.go` line 1543 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtlTJbDlbcA1Xr6MLr:-NbuVC3gBv8Fz8izkVlF:b6qxt0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1543)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same question here, and also in the `break` below. I wonder if this logic was assigning `retErr = res.Err()` before and then it changed.
</blockquote></details>

it's just a bit confusing. this case as well as the above ones do result in the error being shown. since we break from the loop, and `res.Err()` is non-nil, that is the error that is sent back to the user

i can add comments about this.
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 5351 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnUglBB2c26Is1F98:-NbuCPkmD0NfNBwF9Lm6:bd30ifx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/logictest/testdata/logic_test/information_schema#L5351)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: the indentation looks off here and in a few other logic test files. Does `--rewrite` fix that?
</blockquote></details>

good catch
___
*[`pkg/sql/tests/read_committed_test.go` line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtoUR9DZUE_oa2TXCj:-NbuCATKEUq7GX0EWnhz:b-nuf5gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1895f45d440541bbd1d4134afa35c39ef7131b/pkg/sql/tests/read_committed_test.go#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: what does "serializable" refer to here?
</blockquote></details>

just that a serializable transaction was the one writing. i'll rename it to `waitBeforeExternalWriteToC` which might be more clear.
___
*[`pkg/sql/tests/read_committed_test.go` line 76 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtofEiEg2SjghR9-GE:-NbuBqPI6T4sXvInTqMy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1895f45d440541bbd1d4134afa35c39ef7131b/pkg/sql/tests/read_committed_test.go#L76)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider calling out here that `k` is not the primary key, so locking won't be pushed into the initial scan of UPDATE statements.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @nvanbenschoten, and @rhu713)

___
*[`pkg/sql/conn_executor_exec.go` line 1063 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkCG-29dA_EcF_0cb:-NbuXLU16O3mnaFlKDwU:b-cfcgmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1063)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

yeah i'd considered doing it this way, and i'm fine with either approach
</blockquote></details>

on second thought, this converting to event logic may not be needed at all. these errors would all be with "internal" things that the state machine wouldn't really know how to handle anyway, so it's better to treat them as a go error rather than an error event.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @nvanbenschoten, and @rhu713)

___
*[`pkg/sql/conn_executor_exec.go` line 1514 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtkuoxBilZ5wmqkHpw:-NbuZoPK0_R8E-oACkQy:b-tunb8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1514)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

yeah good call
</blockquote></details>

actually, i'll move my comment from the callsite since it fits better here - i extracted this logic into a function after Michael's review but didn't move the comment.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @nvanbenschoten, and @rhu713)

___
*[`pkg/sql/conn_io.go` line 849 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtmlBV-BVy4hR4Oayz:-NburSuIDWbOKcSk-iBi:b7ws3by) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_io.go#L849)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this return a `CmdPos`? And should `TruncateBufferedResults` accept one?
</blockquote></details>

no - `CmdPos` refers to the position in the buffer of commands from the client

for example, `flushInfo.lastFlushed` is described as:
```
// lastFlushed indicates the highest command for which results have been
// flushed. The command may have further results in the buffer that haven't  
// been flushed.
```

i think my naming choice of `pos` added confusion here, so i changed it
___
*[`pkg/sql/conn_io.go` line 852 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtmCInAw2VzT2O3LgJ:-Nbv7tHuDTsrlmw4aYB3:bn1emn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_io.go#L852)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If no results were truncated because no results were added since the call to `BufferedResultsLen`, do we want to return `true` or `false`? In other words, what should `TruncateBufferedResults(BufferedResultsLen())` return?

Given the use case for this method, I'd think we would want it to return true, but I think it will currently return false.
</blockquote></details>

`TruncateBufferedResults(BufferedResultsLen())` was already returning true

but maybe we want `TruncateBufferedResults(BufferedResultsLen() + 1)` to also return true. i'll make that change
___
*[`pkg/sql/vars.go` line 2034 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107044#gh-1267330389:-NburPgx0UpPNrKp0wW6:b96dj8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c309844935d4f6a8e850f4af307fb62deda9836/pkg/sql/vars.go#L2034)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

bikeshedding: I understand that "_transactions" refers to the fact that the whole transaction is at read-committed isolation, but find it a little confusing for this setting. I think it would be clearer to leave it off or change it to "_statements".
</blockquote></details>

i removed `_transactions` from the name
___
*[`pkg/sql/pgwire/command_result.go` line 273 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbtnPYLDdTrNlebR-1j:-Nbv803u6wtarqRZJu29:bw9ou2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/pgwire/command_result.go#L273)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we add a unit test for this?
</blockquote></details>

great point; adding


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
You should be able to rebase this on master now.

> Reviewed 10 of 13 files at r17, 5 of 5 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-:-NbypNX48S6I3zK-Pag7:b-4d690w)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @rafiss, and @rhu713)

___
*[`pkg/sql/conn_executor_exec.go` line 1523 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nbtl0PG0taTWWoDzq2l:-Nbyppu-1URchHJ-3TuL:boxl1qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0363ac26c392870e2f238a7a53efe0d0bdfee9c2/pkg/sql/conn_executor_exec.go#L1523)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one thought i had is if it would make sense to reuse the `autoRetryCounter` that we already use the for the conn_executor state machine retries. it would be a slight abuse of the variable, but it actually doesn't seem incorrect to me, and it would be the quickest way to plug into any existing observability that shows retries.
</blockquote></details>

This makes sense to me. I don't think we want to consider this a "transaction retry" and so we will also want to introduce some "statement retry" specific observability, but in existing cases where we are willing to broaden the definitions (e.g. "auto-retry"), it feels reasonable to group this in.
___
*[`pkg/sql/conn_executor_exec.go` line 1560 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbyopqC4obsblBsl1WK:-NbyopqC4obsblBsl1WL:b-gp8g9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c37b4c57cccb22eaf93aefa38320b45428a45b0/pkg/sql/conn_executor_exec.go#L1560)):*
> ```Go
> 			// flushed and sent back to the client already. In that case, we can't
> 			// retry the statement.
> 			res.SetError(errors.Wrapf(
> ```

Do we have a test that exercises this case?
___
*[`pkg/sql/tests/read_committed_test.go` line 58 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbyogJy6zERPphcmUkl:-NbyogJy6zERPphcmUkm:bpvrqmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c37b4c57cccb22eaf93aefa38320b45428a45b0/pkg/sql/tests/read_committed_test.go#L58)):*
> ```Go
> 				// the second read committed write begins, the read committed scans
> 				// will have finished.
> 				if readCommittedWriteCount.Load() == 2 {
> ```

Instead of re-loading, do we want to use the result of the `Add`? Right now, this is racy.


<!-- Sent from Reviewable.io -->

rafiss committed:
sql: add a test for not doing retries if results were sent

This test ensures that we do not do per-statement retries for READ
COMMITTED transactions if results were already sent to the client.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @nvanbenschoten, and @rhu713)

___
*[`pkg/sql/conn_executor_exec.go` line 1560 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbyopqC4obsblBsl1WK:-Nbzpw_Y16NwGvmVXyca:bhnwd5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c37b4c57cccb22eaf93aefa38320b45428a45b0/pkg/sql/conn_executor_exec.go#L1560)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we have a test that exercises this case?
</blockquote></details>

added a test
___
*[`pkg/sql/tests/read_committed_test.go` line 58 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-NbyogJy6zERPphcmUkl:-Nbzpodd023FTP_52qsc:bsbfpwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c37b4c57cccb22eaf93aefa38320b45428a45b0/pkg/sql/tests/read_committed_test.go#L58)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of re-loading, do we want to use the result of the `Add`? Right now, this is racy.
</blockquote></details>

good point; fixed


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

We might also want one person from SQL to sign off on the changes as well.

> Reviewed 21 of 25 files at r20, 2 of 2 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-:-Nc-2eQO1qSXSV868CwD:bpqxdnc)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, @rafiss, and @rhu713)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice! :champagne:

I left some comments about tests that you can ignore.

> Reviewed 10 of 13 files at r17, 25 of 25 files at r19, 2 of 25 files at r20, 1 of 2 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-:-Nc-J3-t80T3WbANkAUu:bp57vfp)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @rafiss, and @rhu713)

___
*[`pkg/sql/conn_executor_test.go` line 1442 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nc-EoUJ-cL3WsN1bVVo:-Nc-EoUJ-cL3WsN1bVVp:bhxil1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ac84c4f032543cbc6d4b7521fe364106b312a6d/pkg/sql/conn_executor_test.go#L1442)):*
> ```Go
> 	params := base.TestServerArgs{}
> 
> 	var readCommittedStmtRetries int
> ```

Does this also need to be an atomic?
___
*[`pkg/sql/tests/read_committed_test.go` line 49 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nc-GL5B4MgX5Gt6IZGF:-Nc-GF1s9Cpu-kXXtofp:b-d0xn89) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ac84c4f032543cbc6d4b7521fe364106b312a6d/pkg/sql/tests/read_committed_test.go#L49)):*
> ```Go
> 
> 	filterFunc := func(ctx context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
> 		if ba.Txn == nil || ba.Txn.IsoLevel != isolation.ReadCommitted {
> ```

I'm worried that if we ever add internal queries running at ReadCommitted isolation they will cause this test to flake because we're looking for any `Put` at RC isolation. I don't think we need to make it bulletproof now, but maybe a comment would help someone fix the hypothetical flake one day.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107044)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @michae2, and @rhu713)

___
*[`pkg/sql/conn_executor_test.go` line 1442 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nc-EoUJ-cL3WsN1bVVo:-Nc-NJYFCv92opvWFKz4:bsag21q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ac84c4f032543cbc6d4b7521fe364106b312a6d/pkg/sql/conn_executor_test.go#L1442)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does this also need to be an atomic?
</blockquote></details>

hmm it might be safer. will do
___
*[`pkg/sql/tests/read_committed_test.go` line 49 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/107044#-Nc-GL5B4MgX5Gt6IZGF:-Nc-NbRf4GphbMQYVgNz:b-l04o8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ac84c4f032543cbc6d4b7521fe364106b312a6d/pkg/sql/tests/read_committed_test.go#L49)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm worried that if we ever add internal queries running at ReadCommitted isolation they will cause this test to flake because we're looking for any `Put` at RC isolation. I don't think we need to make it bulletproof now, but maybe a comment would help someone fix the hypothetical flake one day.
</blockquote></details>

i think i can easily change this so it only counts writes to the table in this test


<!-- Sent from Reviewable.io -->

rafiss commented:
tftrs!

bors r+

rafiss commented:
bors r+

rafiss commented:
[flake](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_JaneDockerImage/11354439?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true) was:
```
   > [internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:latest:
  ------
  Dockerfile:1
  --------------------
     1 | >>> FROM registry.access.redhat.com/ubi8/ubi-minimal
     2 |     ARG fips_enabled
     3 |
  --------------------
  ERROR: failed to solve: registry.access.redhat.com/ubi8/ubi-minimal: failed to do request: Head "https://registry.access.redhat.com/v2/ubi8/ubi-minimal/manifests/latest": read tcp 10.142.0.162:46278->23.47.144.132:443: read: connection reset by peer
  + remove_files_on_exit
```

bors r+

rafiss commented:
bors r+
